File: eval/oracle/android/src/test/kotlin/notes/StubNotesApi.kt
package notes

import kotlin.math.ceil

/** In-memory NotesApi for the oracle: deterministic search/tag/pagination so the
 *  view-model behaviour is the only thing under test. */
class StubNotesApi(
    private val all: List<Note>,
    private val validCredentials: Pair<String, String>? = "a@b.com" to "pw",
    private val failListWith: NotesError? = null,
) : NotesApi {
    var listCallCount = 0
        private set

    override suspend fun login(email: String, password: String): Session {
        val creds = validCredentials
        if (creds != null && creds.first == email && creds.second == password) {
            return Session("tok")
        }
        throw NotesError.Server("Invalid credentials")
    }

    override suspend fun listNotes(query: NotesQuery): NotesPage {
        listCallCount++
        failListWith?.let { throw it }

        var filtered = all
        query.search?.takeIf { it.isNotEmpty() }?.let { s ->
            filtered = filtered.filter { it.title.contains(s, ignoreCase = true) }
        }
        query.tag?.let { t -> filtered = filtered.filter { it.tags.contains(t) } }

        val total = filtered.size
        val size = maxOf(1, query.pageSize)
        val totalPages = maxOf(1, ceil(total.toDouble() / size).toInt())
        val start = (query.page - 1) * size
        val slice = filtered.drop(maxOf(0, start)).take(size)
        return NotesPage(slice, query.page, totalPages, total)
    }
}

File: eval/oracle/ios/Tests/NotesFeatureOracleTests/StubNotesAPI.swift
import Foundation
import NotesFeature

/// In-memory NotesAPI used by the oracle. Applies search, tag, and pagination
/// deterministically so view-model behaviour is the only thing under test.
final class StubNotesAPI: NotesAPI, @unchecked Sendable {
    let all: [Note]
    let validCredentials: (email: String, password: String)?
    let failListWith: NotesError?
    private(set) var listCallCount = 0

    init(all: [Note],
         validCredentials: (email: String, password: String)? = ("a@b.com", "pw"),
         failListWith: NotesError? = nil) {
        self.all = all
        self.validCredentials = validCredentials
        self.failListWith = failListWith
    }

    func login(email: String, password: String) async throws -> Session {
        if let c = validCredentials, c.email == email, c.password == password {
            return Session(token: "tok")
        }
        throw NotesError.server("Invalid credentials")
    }

    func listNotes(_ query: NotesQuery) async throws -> NotesPage {
        listCallCount += 1
        if let failure = failListWith { throw failure }

        var filtered = all
        if let s = query.search, !s.isEmpty {
            filtered = filtered.filter { $0.title.range(of: s, options: .caseInsensitive) != nil }
        }
        if let t = query.tag {
            filtered = filtered.filter { $0.tags.contains(t) }
        }

        let total = filtered.count
        let size = max(1, query.pageSize)
        let totalPages = max(1, Int(ceil(Double(total) / Double(size))))
        let start = (query.page - 1) * size
        let slice = Array(filtered.dropFirst(max(0, start)).prefix(size))
        return NotesPage(notes: slice, page: query.page, totalPages: totalPages, totalCount: total)
    }
}
